piper/mousemap.py

```
@@ -0,0 +1,390 @@
+    should also be given to the `do_add` method. See `do_size_allocate` and
+    https://github.com/libratbag/libratbag/blob/master/data/README.md for more
+    information.
+    """
+
+    __gtype_name__ = "MouseMap"
+
+    __gproperties__ = {
+        "spacing": (int,
+                    "spacing",
+                    "The amount of space between children and the SVG leaders",
+                    0, GLib.MAXINT, 0,
+                    GObject.PARAM_READABLE),
+    }
+
+    def __init__(self, layer, ratbagd_device, spacing=10, *args, **kwargs):
+        """Instantiates a new MouseMap.
+
+        @param layer The SVG layer whose leaders to draw, according to librsvg
+                     so e.g. `#layer1`.
+        @param ratbagd_device The device that should be mapped, as
+                              ratbagd.RatbagdDevice
+        @param spacing The spacing to place between the SVG's leaders and the
+                       widgets, as int
+
+        @raises ValueError when an argument is invalid, or when the given device
+                           has no image registered.
+        @raises GLib.Error when the SVG cannot be loaded.
+        """
+        if layer is None:
+            raise ValueError("Layer cannot be None")
+        if ratbagd_device is None:
+            raise ValueError("Device cannot be None")
+        if not os.path.isfile(ratbagd_device.svg_path):
+            raise ValueError("Device has no image or its path is invalid")
+
+        Gtk.Container.__init__(self, *args, **kwargs)
+        self.set_has_window(False)
+
+        self.spacing = spacing
+        self._layer = layer
+        self._device = ratbagd_device
+        self._children = []
+        self._highlight_element = None
+
+        self._handle = Rsvg.Handle.new_from_file(ratbagd_device.svg_path)
+        self._svg_data = etree.parse(ratbagd_device.svg_path)
+
+    def do_add(self, widget):
+        """Not implemented, use `add(widget, svg_id)` instead."""
+        pass
```

should we throw an exception here? seems like a programming error if that is ever called